### PR TITLE
awful.tag: Add `gap_single_client` property and user changeable defau…

### PR DESCRIPTION
**Pull request by Akulen**
This is a reimplementation of [https://github.com/awesomeWM/awesome/pull/860], with `gap_count` being renamed to `gap_single_client` and changed to a boolean (true meaning that useless gaps are allowed even when only one client is present)

### CODE REVIEW

File: lib/awful/layout/init.lua [hunk 1]
```diff
@@ -137,7 +137,10 @@ function layout.parameters(t, screen)
 
     local p = {}
 
-    local useless_gap = t and t.gap or 0
+    local clients           = client.tiled(screen)
+    local gap_single_client = t and t.gap_single_client or true
```
Comment (actionless) [inline lib/awful/layout/init.lua]:
```
$ lua
Lua 5.3.3  Copyright (C) 1994-2016 Lua.org, PUC-Rio
> true and false or true
true
```

**Comment by actionless** [comment]
also i think it could useful to add for a single client some logic similar to the one presented in this code sample:
https://github.com/awesomeWM/awesome/issues/826

**Comment by Elv13** [comment]
I changed my mind a little since I made the original versions. The `.defaults` variables should become hardcoded again, but each should have a `@beautiful` property. This way themes can set their own `master_width_factor`

**Comment by Akulen** [comment]
I've fixed the value attribution for `gap_single_client` in init.lua and each of the defaults variable are overriden by their beautiful counterparts. I dont know it the defaults value should be hardcoded (maybe just not documented)

File: lib/awful/tag.lua [hunk 2]
```diff
@@ -41,6 +41,41 @@ data.tags = setmetatable({}, { __mode = 'k' })
 tag.history = {}
 tag.history.limit = 20
 
+-- Default values
+tag.defaults = {}
+--- The gap between clients (in points).
+-- @tfield[opt=0] integer tag.defaults.gap
+-- @tparam[opt=0] integer gap
+-- @see gap
+-- @see gap_single_client
```
Comment (Elv13) [inline lib/awful/tag.lua]: This will need some empty lines between the block

**Comment by Elv13** [comment]
> I dont know it the defaults value should be hardcoded (maybe just not documented)

I was thinking about turning the vars into `beautiful` theme variables. Then hardcoding the definitive fallback.

In some other unrelated issue, I discussed the possibility of adding hardcoded fallbacks to all theme variables so incomplete themes become supported again. I see this feature as something a theme might want to set.

**Comment by Akulen** [comment]
> I was thinking about turning the vars into beautiful theme variables. Then hardcoding the definitive fallback.

I've already added a beautiful counterpart to all the defaults variables, but i was thinking that they could still be centralized in a defaults array (maybe just in `local defaults = {}` ...) to make it easier to see and edit the default values if ever needed.

**Comment by actionless** [comment]
>> I've already added a beautiful counterpart to all the defaults variables, but i was thinking that they could still be centralized in a defaults array (maybe just in local defaults = {} ...) to make it easier to see and edit the default values if ever needed.

good idea but i would suggest using `awful.layout.defaults = {...}` or smth like that instead of `local defaults = {...}` since such approach is already used in some awesome modules (but they're documented, so i would recommend to document those as well)

**Comment by Akulen** [comment]
>  i would suggest using `awful.layout.defaults = {...}`

Yeah, it seems better than to use a local array, but these are properties for `awful.tag`, so i don't think `awful.layout.defaults` is the correct location. What I think should be used is `awful.tag.defaults`, but that's already the case in my PR and i've documented them as well

File: lib/awful/layout/init.lua [hunk 3]
```diff
@@ -137,7 +137,13 @@ function layout.parameters(t, screen)
 
     local p = {}
 
-    local useless_gap = t and t.gap or 0
+    local clients           = client.tiled(screen)
+    local gap_single_client = true
+    if(t and not (t.gap_single_client == nil)) then
```
Comment (Elv13) [inline lib/awful/layout/init.lua]: Please use `~=` and add a new line before and after for readability

File: lib/awful/tag.lua [hunk 4]
```diff
@@ -41,6 +41,21 @@ data.tags = setmetatable({}, { __mode = 'k' })
 tag.history = {}
 tag.history.limit = 20
 
+-- Default values
+local defaults = {}
+-- The gap between clients (in points).
```
Comment (Elv13) [inline lib/awful/tag.lua]: Please add new lines between the variables for consistency with other files

File: lib/awful/tag.lua [hunk 5]
```diff
@@ -570,7 +593,7 @@ function tag.object.set_master_width_factor(t, mwfact)
 end
 
 function tag.object.get_master_width_factor(t)
-    return tag.getproperty(t, "master_width_factor") or 0.5
+    return tag.getproperty(t, "master_width_factor") or beautiful.master_width_factor or defaults.master_width_factor
```
Comment (Elv13) [inline lib/awful/tag.lua]: we have a soft 80 line policy (most of the time)

```lua
return tag.getproperty(t, "master_width_factor")
    or beautiful.master_width_factor
    or defaults.master_width_factor
```

File: lib/awful/tag.lua [hunk 6]
```diff
@@ -835,7 +892,7 @@ end
 -- @param string "expand" or "master_width_factor"
 
 function tag.object.get_master_fill_policy(t)
-    return tag.getproperty(t, "master_fill_policy") or "expand"
+    return tag.getproperty(t, "master_fill_policy") or beautiful.master_fill_policy or defaults.master_fill_policy
```
Comment (Elv13) [inline lib/awful/tag.lua]: see the previous comment (line length)

File: lib/awful/tag.lua [hunk 7]
```diff
@@ -1005,7 +1074,7 @@ function tag.object.set_column_count(t, ncol)
 end
 
 function tag.object.get_column_count(t)
-    return tag.getproperty(t, "column_count") or 1
+    return tag.getproperty(t, "column_count") or beautiful.column_count or defaults.column_count
```
Comment (Elv13) [inline lib/awful/tag.lua]: same

File: lib/awful/tag.lua [hunk 8]
```diff
@@ -878,9 +935,15 @@ function tag.getmfpol(t)
     util.deprecate("Use t.master_fill_policy instead of awful.tag.getmfpol")
 
     t = t or ascreen.focused().selected_tag
-    return tag.getproperty(t, "master_fill_policy") or "expand"
+    return tag.getproperty(t, "master_fill_policy") or beautiful.master_fill_policy or defaults.master_fill_policy
```
Comment (Elv13) [inline lib/awful/tag.lua]: this one too

**Comment by Elv13** [comment]
👍

File: lib/awful/tag.lua [hunk 9]
```diff
@@ -41,6 +41,27 @@ data.tags = setmetatable({}, { __mode = 'k' })
 tag.history = {}
 tag.history.limit = 20
 
+-- Default values
+local defaults = {}
```
Comment (actionless) [inline lib/awful/tag.lua]: but `defaults` is still local

Comment (Elv13) [inline lib/awful/tag.lua]: Yes, I request that change on IRC. Sorry about that, but I don't think having 2 places to change the default is a good idea. You can set beautiful variables in rc.lua if you want. I didn't wanted to have to deal with the confusion having an exported default table would ensue.

**Pull request merged**